src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs

```diff
@@ -358,6 +359,43 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
                 res.SetHandle(acceptedFd);
                 return res;
             }
+
+            internal unsafe void TryUnblockSocket(bool abortive)
+            {
+                // Calling 'close' on a socket that has pending blocking calls (e.g. recv, send, accept, ...)
+                // may block indefinitly. This is a best effort attempt to not get blocked and make those operations return.
+                // We need to ensure we keep the expected TCP behavior that is observed by the socket peer (FIN vs RST close).
+                // What we do here isn't specified by POSIX, it works for Linux.
+
+                // Don't touch sockets which don't have the CLOEXEC flag set. These may be shared
+                // with other processes and we want to avoid disconnecting them.
+                int fdFlags = Interop.Sys.Fcntl.GetFD(this);
+                if (fdFlags == 0)
+                {
+                    return;
+                }
+
+                var linger = new Interop.Sys.LingerOption();
+                if (abortive ||
+                    (Interop.Sys.GetLingerOption(this, &linger) == Interop.Error.SUCCESS && linger.OnOff != 0))
+                {
+                    if (abortive || linger.Seconds == 0)
```

Comment:
Can you comment why the linger settings are incorporated into these checks? e.g. if linger is on with > 0 seconds, we don't do anything?

Reply:
The operation has an immediate effect so I am not handling linger timeout > 0 atm.